### PR TITLE
throttler: First version of adaptive throttling.

### PR DESCRIPTION
**Pull request by michael-berlin**
@alainjobart 

Alain, please ignore all but the last 4 commits. The other commits are a pending pull request.

For the review I suggest you review things at a high level first e.g. read each added struct and then review one commit at a time.

Missing/Broken things which I'll address in future commits:

- vtworker unit tests are hanging.
- Not all debug statements may be removed.
- history data structures are not bounded.
- Support to ignore updates from the N slowest replicas.
- in increase state: Wait until the same replica send an update again before we increase the rate again.
- memory.go: Bad values must age out after time.
- vtctl: Add command to change module config on the fly.
- Add vars which allow us to monitor the state of the module.
- Write unit tests for max_replication_lag_module.go.

### CODE REVIEW

**Comment by alainjobart** [comment]
It seems reviewable cannot get just the last 4 commits, so I had to do it with github. Sorry for the emails.

File: go/vt/worker/split_clone.go [hunk 1]
```diff
@@ -965,3 +1003,67 @@ func (scw *SplitCloneWorker) getSourceSchema(ctx context.Context, tablet *topoda
...
 	return sourceSchemaDefinition, nil
 }
+
+// StatsUpdate receives replication lag updates for each destination master
+// and forwards them to the respective throttler instance.
+// It is part of the discovery.HealthCheckStatsListener interface.
+func (scw *SplitCloneWorker) StatsUpdate(ts *discovery.TabletStats) {
+	scw.tsc.StatsUpdate(ts)
+
+	// Ignore if not MASTER or REPLICA.
```
Comment (alainjobart) [inline go/vt/worker/split_clone.go]: You need to handle master in a different way, right?

File: go/vt/worker/split_clone.go [hunk 2]
```diff
@@ -965,3 +1003,67 @@ func (scw *SplitCloneWorker) getSourceSchema(ctx context.Context, tablet *topoda
...
+	}
+
+	// Lock throttlers mutex to avoid that this method (and the called method
+	// Throttler.RecordReplicationLag()) races with closeThrottlers() (which calls
+	// Throttler.Close()).
+	scw.throttlersMu.Lock()
+	defer scw.throttlersMu.Unlock()
+
+	t := scw.getThrottlerLocked(ts.Target.Keyspace, ts.Target.Shard)
+	// TODO(mberlin): Track from which replica the lag came as well. Necessary to ignore the N slowest replicas.
```
Comment (alainjobart) [inline go/vt/worker/split_clone.go]: Yep you probably need to keep the list yourself here.

File: go/vt/worker/split_clone.go [hunk 3]
```diff
@@ -965,3 +1003,67 @@ func (scw *SplitCloneWorker) getSourceSchema(ctx context.Context, tablet *topoda
 	}
 	return sourceSchemaDefinition, nil
 }
+
+// StatsUpdate receives replication lag updates for each destination master
+// and forwards them to the respective throttler instance.
+// It is part of the discovery.HealthCheckStatsListener interface.
+func (scw *SplitCloneWorker) StatsUpdate(ts *discovery.TabletStats) {
```
Comment (alainjobart) [inline go/vt/worker/split_clone.go]: If ts.Up is false, the tablet is gone from serving. You should not consider it any more. Overall, you need to keep a map from ts.Key to *discovery.TabletStats, and have a function that returns the replication lag you want from that map (second-to-smallest maybe?).

File: go/vt/throttler/demo/throttler_demo.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+//
+// This demo simulates a client (writer), a master and a replica.
+// The client writes to the master which in turn replicas everything to the
+// replica.
+// The replica measures its replication lag via the timestamp which is part of
+// each message.
+// While the master has no rate limit, the replica is limited to
+// --rate (see below) transactions/second. The client runs the resharding
+// throttler which tries to throttle the client based on the observed
+// replication lag.
```
Comment (alainjobart) [inline go/vt/throttler/demo/throttler_demo.go]: Would be nice to be able to inject a few seconds delay at some point in there, see how we react to the spike.

File: go/vt/discovery/tablet_stats_cache.go [hunk 5]
```diff
@@ -48,14 +48,31 @@ type tabletStatsCacheEntry struct {
...
 	}
 
-	// We need to set sendDownEvents=true to get the deletes from the map
-	// upon type change.
-	hc.SetListener(tc, true /*sendDownEvents*/)
+	if setListener {
+		// We need to set sendDownEvents=true to get the deletes from the map
+		// upon type change.
+		hc.SetListener(tc, true /*sendDownEvents*/)
+	}
```
Comment (alainjobart) [inline go/vt/discovery/tablet_stats_cache.go]: Maybe just have a single constructor without the HC, and add a:
func (tc *TabletStatsCache) ListenToHealthCheck(hc HealthCheck) {
   hc.SetListener(tc, true /*sendDownEvents*/)
}
and call it where necessary. Might be simpler.

**Comment by alainjobart** [comment]
OK this is hard to review as is.  However, this is limited to throttler module and some worker code. I have a few comments as is. I think the best course of action is for you to fix the tests, and my couple comments (well, not the comments about better handling of StatsUpdate, that will be too big), to get travis to pass, then submit this. I'll review it more after that. But you can also incrementally fix things from that point on. Please do not add new features (like better handling of StatsUpdate) in this PR, first fix and submit this one.

My main overall concern is a situation where we set a 'bad' rate when transactions/s is low because of big rows, and then we can't go past that bad rate when the rows get small. I think you have it in there as forgetting bad rates after a while, that may do it...

**Comment by alainjobart** [comment]
OK I agree with you, revert the change I proposed for the healthcheck listener. New code is not pretty. I think I need to rework this stuff independently (imo, the listener should be created before and freed after the healthcheck object, not the other way around, which leads to this crazy code. I'll change that myself later). So just revert https://github.com/youtube/vitess/commit/e02a6aa

The Master fix looks good. Do you also want to ignore ts.Up = false events for now? Or just fix it as part of the StatsUpdate change, that's fine too...

LGTM after these are addressed.

**Comment by michael-berlin** [comment]
Tests for this PR are passing except for the broken unit_test_race.

Therefore I'm merging this now.

This should not affect people who are trying to run SplitClone because the adapative throttling is still disabled there. See the `NewThrottler()` call in `split_clone.go`.

**Pull request merged**